clang-tools-extra/clang-tidy/bugprone/NullCheckAfterDereferenceCheck.cpp

```diff
@@ -0,0 +1,125 @@
+
+static std::optional<ReverseNullDiagnoser::ResultType>
+analyzeFunction(const FunctionDecl &FuncDecl, ASTContext &ASTCtx) {
+  using dataflow::ControlFlowContext;
+  using dataflow::DataflowAnalysisState;
+  using llvm::Expected;
+
+  if (FuncDecl.getBody() == nullptr) {
+    return std::nullopt;
+  }
+
+  Expected<ControlFlowContext> Context =
+      ControlFlowContext::build(FuncDecl, *FuncDecl.getBody(), ASTCtx);
+  if (!Context)
+    return std::nullopt;
+
+  dataflow::DataflowAnalysisContext AnalysisContext(
+      std::make_unique<dataflow::WatchedLiteralsSolver>());
+  dataflow::Environment Env(AnalysisContext, FuncDecl);
+  NullCheckAfterDereferenceModel Analysis(ASTCtx);
+  Analysis.initializeFunctionParameters(*Context, Env);
+  ReverseNullDiagnoser Diagnoser;
+  ReverseNullDiagnoser::ResultType Diagnostics;
+  Expected<std::vector<std::optional<
+      DataflowAnalysisState<NullCheckAfterDereferenceModel::Lattice>>>>
+      BlockToOutputState = dataflow::runDataflowAnalysis(
+          *Context, Analysis, Env,
+          [&ASTCtx, &Diagnoser, &Diagnostics](
+              const CFGElement &Elt,
+              const DataflowAnalysisState<
+                  NullCheckAfterDereferenceModel::Lattice> &State) mutable {
+            auto EltDiagnostics = Diagnoser.diagnose(ASTCtx, &Elt, State.Env);
+            llvm::move(EltDiagnostics.first,
+                       std::back_inserter(Diagnostics.first));
+            llvm::move(EltDiagnostics.second,
+                       std::back_inserter(Diagnostics.second));
+          });
+
+  // FIXME: Port this over into UncheckedOptionalAccessCheck
+  if (llvm::Error E = BlockToOutputState.takeError()) {
+    llvm::dbgs() << "Dataflow analysis failed: " << llvm::toString(std::move(E))
+                 << ".\n";
+    return std::nullopt;
+  }
+
+  return Diagnostics;
+}
+
+void NullCheckAfterDereferenceCheck::registerMatchers(MatchFinder *Finder) {
+  using namespace ast_matchers;
+
+  auto hasPointerValue =
+      hasDescendant(NullCheckAfterDereferenceModel::ptrValueMatcher());
+  Finder->addMatcher(
+      decl(anyOf(functionDecl(unless(isExpansionInSystemHeader()),
+                              // FIXME: Remove the filter below when lambdas are
+                              // well supported by the check.
+                              unless(hasDeclContext(cxxRecordDecl(isLambda()))),
+                              hasBody(hasPointerValue)),
+                 cxxConstructorDecl(hasAnyConstructorInitializer(
+                     withInitializer(hasPointerValue)))))
+          .bind(FuncID),
+      this);
+}
+
+void NullCheckAfterDereferenceCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  // llvm::errs() << "Starting result parse\n";
+  if (Result.SourceManager->getDiagnostics().hasUncompilableErrorOccurred())
+    return;
+
+  const auto *FuncDecl = Result.Nodes.getNodeAs<FunctionDecl>(FuncID);
+  if (FuncDecl->isTemplated())
```

Comment:
Conventionally, there is an assert here to ensure that if someone breaks the matcher(s) it is caught without having to rely on U.B.

```suggestion
  const auto *FuncDecl = Result.Nodes.getNodeAs<FunctionDecl>(FuncID);
  assert(FuncDecl && "Matcher failed."); // Or "bogus matcher" or something like that.
  if (FuncDecl->isTemplated())
```

Reply:
Done.